### ISSUE TITLE
"gocql: no hosts available in the pool" cannot recover without restart

### ISSUE THREAD

**Issue by zhixinwen** [gocql/gocql#915]
We had a cluster degrades due to increased traffic load, and on the client side we see "gocql: no hosts available in the pool" error for ~80% of the requests.

The error rate is very consistent, and the C* cluster is healthy except the load. We try to reduce the load on the cluster, but the error is still consistent. It lasts for two hours.

We restart the client, and the errors go away immediately.

Also one thing interesting is that after we restart the gocql client, we also see an immediate drop on coordinator latency and traffic on the C* cluster (Yes, client restarts first, then the latency and traffic on C* drops, not the other way around). It may be due to upstream users stop retry after we restart the server and no longer see "gocql: no hosts available in the pool", but we are not sure about the root cause yet.

**Comment by zhixinwen** [comment]
After we close a connection due to too many timeouts, when would the connection be made again? 

I do see gocql wait for event from control connection to reconnect. But if no node is perceived as down by the control node, and it never emits a 'node up' event, would gocql try to recreate the connections?

**Comment by zhixinwen** [comment]
@Zariel gocql closes a connection when it takes too long to respond, when would the connection be reconnected?

I may miss some of the code, looks like it will only reconnect when it receives "Node Up" event?

**Comment by Zariel** [comment]
It should be periodically try to dial nodes that are down, if you can reproduce and build with `gocql_debug` you should see more info about the pool state.

**Comment by zhixinwen** [comment]
I think the root cause of the problem here is:

1. When all connections to a hosts are closed, the host is not marked as down, so ReconnectInterval would not help
2. HandleError in hostConnPool would try to refill the connection pool when a connection is closed, but in an outage it is very likely to fail. Moreover it would not retry afterward. As a result, when the cluster recovers after all the connections are closed, gocql would not attempt to rebuild the connection pool

To fix this, we need to mark the host as DOWN, when we close all the connections on that host.

**Comment by balta2ar** [comment]
I simulated network connectivity issues using iptables:
```
sudo iptables -I OUTPUT -p tcp -d 127.0.0.1 --dport 9042 -j DROP
```
After that upon queries I start to receive this error message: `gocql: no response received from cassandra within timeout period`
And soon after it this message: `gocql: no hosts available in the pool`

After deleting the rule (`sudo iptables -D OUTPUT 1`) gocql indeed does not recover automatically.

**Comment by Zariel** [comment]
Running a single node of Cassandra (on mac so excuse lack of iptables), able to query host as expected when its running. Kill -9 the process, gocql receives no host down notification and discovers the control connection is dead and marks the host down. Restarting Cassandra the driver reconnects and queries can continue. Repeated with STOP/CONT and had the same results.

Can you rerun your test with `gocql_debug` build tag

**Comment by zhixinwen** [comment]
I think `kill -9` is not a accurate simulation, simulation of @balta2ar seems to be more similar to the issue I see.

**Comment by Zariel** [comment]
Do you know of a way to replicate on OSX or reliably for regression tests?

**Comment by robdefeo** [comment]
When this error occurs I have had to cycle the container as there is no good way to reconnect after this event. There must be a better way but this error has caused so many problems

**Comment by engin** [comment]
seems that setting `ClusterConfig.ReconnectInterval` > 0 has no affect on this.

**Comment by shettyh** [comment]
+1

Any fixes for this issue ?

**Comment by steebchen** [comment]
Constantly seeing this issue on single-node Cassandra cluster. `cqlsh` works fine 100% of the time while gocql seems to take a lot of time connecting to the cluster, often resulting in "gocql: no hosts available in the pool" error.

**Comment by rlk833** [comment]
We have this problem too. Have had it since at least Oct of 2018. I had to put in special code to detect this and cause the application to be recycled. Our apps call gocql under very heavy load (millions an hour requests). For a long time the recycle occurred once a month or so. But the loads have increased recently and it is now up to every days or even shorter. We have 4 instances running. It is not always all 4 instances at the same time. Out of nowhere with no proceeding errors logged we all of a sudden get this in a running instance. The only fix is to restart the instance. Occasionally other instances will shortly run into the problem to and need restarting. If we restart there is no problem at all connecting to the nodes. 

The main problem we have is during the detection and restarting period there is a significant slow down in processing as the instance is not available.

**Comment by clarakosi** [comment]
I ran into a similar issue a couple of days ago and followed @balta2ar [instructions](https://github.com/gocql/gocql/issues/915#issuecomment-325329596) to try and recreate the situation.

And surprisingly gocql reconnected after removing the rule (`sudo iptables -D OUTPUT 1`).

Here are the logs:

```lang=bash
{"msg":"Error reading from storage (gocql: no response received from cassandra within timeout period)","appname":"kask","time":"2020-05-21T18:05:40-04:00","level"
:"ERROR","request_id":"00000000-0000-0000-0000-000000000000"}
{"msg":"gocql: Session.handleNodeDown: 127.0.0.1:9042","appname":"kask","time":"2020-05-21T18:05:47-04:00","level":"WARNING"}
{"msg":"Error reading from storage (gocql: no hosts available in the pool)","appname":"kask","time":"2020-05-21T18:05:47-04:00","level":"ERROR","request_id":"0000
0000-0000-0000-0000-000000000000"}
{"msg":"Error reading from storage (gocql: no hosts available in the pool)","appname":"kask","time":"2020-05-21T18:05:50-04:00","level":"ERROR","request_id":"0000
0000-0000-0000-0000-000000000000"}
{"msg":"Error reading from storage (gocql: no hosts available in the pool)","appname":"kask","time":"2020-05-21T18:05:51-04:00","level":"ERROR","request_id":"0000
0000-0000-0000-0000-000000000000"}
{"msg":"gocql: Session.handleNodeDown: 127.0.0.1:9042","appname":"kask","time":"2020-05-21T18:05:58-04:00","level":"WARNING"}
{"msg":"gocql: Session.handleNodeDown: 127.0.0.1:9042","appname":"kask","time":"2020-05-21T18:06:09-04:00","level":"WARNING"}
{"msg":"gocql: Session.handleNodeDown: 127.0.0.1:9042","appname":"kask","time":"2020-05-21T18:06:20-04:00","level":"WARNING"}
{"msg":"gocql: Session.handleNodeUp: 127.0.0.1:9042","appname":"kask","time":"2020-05-21T18:06:23-04:00","level":"WARNING"}
```

Any other ideas on how to replicate this issue?

**Comment by SubNader** [comment]
Running into this while using gocql to write to AWS Keyspaces.
Any updates on this?

**Comment by mattmassicotte** [comment]
I'm also experiencing this issue extremely frequently with AWS Keyspaces. It appears to be entirely due to state of the process. Other processes can access Keyspaces without issue, and restarting an affected process immediately resolves the issue. I'm up for helping to debug however I can.

**Comment by Draal** [comment]
it seems like it happens because of 
https://github.com/gocql/gocql/blob/390af0fb2915aecc9d18df7afa11501fc2d4c9c5/connectionpool.go#L248
and then we add host back only at https://github.com/gocql/gocql/blob/390af0fb2915aecc9d18df7afa11501fc2d4c9c5/events.go#L187-L187
which if I understand correctly happens only when a new node is starting, but not when a node temporarily lost network connection or due to overload rejected request.

**Comment by martin-sucha** [comment]
There is `pool.addHost` in `reconnectDownedHosts`. `reconnectDownedHosts` periodically adds nodes that are marked as down. However, we set the host state as down only when we get such event from the database, which does not occur if just gocql lost connections (or closed them due to errors).

**Comment by mrwonko** [comment]
I'm trying to reproduce this using docker-compose on Mac. I'm using the `alpine` image for my client, installed iptables via `apk --update add iptables`, then added
```yaml
    cap_add:
      - NET_ADMIN
```
to my docker-compose.yaml so I can use `iptables` inside the container. I have a small Go program that does SELECT requests in a loop.

When I execute `iptables -I OUTPUT -p tcp --dport 9042 -j DROP`, this happens:

```
client_1     | 2021/04/22 15:08:22 SELECTed: "data"
client_1     | 2021/04/22 15:08:23 SELECTed: "data"
client_1     | 2021/04/22 15:08:25 SELECT failed: gocql: no response received from cassandra within timeout period
client_1     | 2021/04/22 15:08:26 gocql: Session.handleNodeDown: 172.22.0.2:9042
client_1     | 2021/04/22 15:08:26 SELECT failed: gocql: connection closed waiting for response
client_1     | 2021/04/22 15:08:26 SELECT failed: gocql: no hosts available in the pool
client_1     | 2021/04/22 15:08:27 SELECT failed: gocql: no hosts available in the pool
client_1     | 2021/04/22 15:08:28 gocql: Session.handleNodeDown: 172.22.0.2:9042
client_1     | 2021/04/22 15:08:28 SELECT failed: gocql: no hosts available in the pool
client_1     | 2021/04/22 15:08:29 SELECT failed: gocql: no hosts available in the pool
```

So far, so good. Occasionally, an automatic reconnect is apparently attempted:

```
client_1     | 2021/04/22 15:09:20 connection failed "172.22.0.2": dial tcp 172.22.0.2:9042: i/o timeout, reconnecting with *gocql.ConstantReconnectionPolicy
```

But when I then remove the rule via `iptables -D OUTPUT 1`, gocql recovers immediately:

```
client_1     | 2021/04/22 15:10:28 SELECT failed: gocql: no hosts available in the pool
client_1     | 2021/04/22 15:10:29 gocql: Session.handleNodeDown: 172.22.0.2:9042
client_1     | 2021/04/22 15:10:29 gocql: Session.handleNodeUp: 172.22.0.2:9042
client_1     | 2021/04/22 15:10:29 SELECTed: "data"
client_1     | 2021/04/22 15:10:30 SELECTed: "data"
```

However I know that the issue is not fixed entirely, because we had an outage due to it recently. Now I don't know how to reproduce it.

**Comment by mrwonko** [comment]
Correction: I had been testing the outdated commit 1418e1105e72. With HEAD (909f2a77f46e) I can reproduce the issue using the instructions above, it looks like the first commit that re-introduced the issue was c80d1b329030. I'll see if I can understand the issue and fix it.

**Comment by mrwonko** [comment]
I'm fairly sure the regression is connected to the removal of [this](https://github.com/gocql/gocql/commit/c80d1b32903033df8db60e88d144b5638799b35a#diff-c5dc4ab579e7639ce45c1d2478d893d08ae82f656dded4fdef9b0a2337717f56L288) call to `session.handleNodeUp()` in `controlConn.setupConn()`, as I can reproduce the issue if I take 1418e11 and just remove that line. But I don't know the code well enough to fix it. I tried to simply add it back in, but that didn't seem to do anything on HEAD, and randomly undoing changes I don't understand is not a good way to fix this.

I'd appreciate if somebody with a better understanding of the code could take over from here.

**Comment by martin-sucha** [comment]
I tried to reproduce using a similar docker-compose setup as above and it seems that 1418e1105e720e09744dd50f9b21aed3911bf434 is much quicker to reconnect than 909f2a77f46e7e16f59959bfce1f7bdcc4bb28c5. The latter reconnects about up to a minute after I delete the iptables rule, looks like because of `ClusterConfig.ReconnectInterval`.

Is that what you observed or did you reproduce a case when the driver did not reconnect at all?

**Comment by maxstepanov** [comment]
Just hit this. Did a rolling restart of C* cluster. Service started throwing a lots of "gocql: no hosts available in the pool".
Had to restart the application.
I am on latest i think
gocql v0.0.0-20210425135552-909f2a77f46e

**Comment by raidancampbell** [comment]
has anyone faced this after #1555 was merged on May 13?  I just encountered it using a version before that patch, and am upgrading now to see if I face it again.  The issue is rare enough for me that it sometimes takes weeks to manifest

**Comment by justinfx** [comment]
This issue is easily reproducible as per my notes on https://github.com/yugabyte/yugabyte-db/issues/10182

I was able to repro this with a Yugabyte operator cluster running on k8s with a simple scale down of all nodes and back up. Problem is exactly as described here and on the other issue. Connection pool indefinitely remains down, and original host names are never reevaluated to pick up new ip addresses.

**Comment by mouglou** [comment]
Hi,

We get the problem too.

Vault: 1.8.2
Cassandra: 3.11.6

When we try to generate some credentials:
`failed to roll back user v_kube_xxxxxx: gocql: no hosts available in the pool`

No error/warn/info in Vault logs about that, no logs on Cassandra too. Firewall ok, netcat ok.
Only the "reset" connection from Vault allow us to go back operational.

https://www.vaultproject.io/api/secret/databases#reset-connection

Any updates ?

**Comment by justinfx** [comment]
So far I have had to do an applications-side hack, to work around the issue, but it still isn't 100%. I'm using the yugabyte forked client, which should be the same upstream problem:
https://github.com/yugabyte/yugabyte-db/issues/10182#issuecomment-993792821

**Comment by player1-github** [comment]
@mouglou Vault 1.8.2 use old version of gocql (https://github.com/hashicorp/vault/blob/main/go.mod#L53). There is PR open to bump the version - https://github.com/hashicorp/vault/pull/12879 and resolve https://github.com/hashicorp/vault/issues/12878 but it was not merged yet. The fix for issue you described was merged in gocql here: https://github.com/gocql/gocql/commit/312a614907c8d2e19f6b14aff16015436cb78423

**Comment by mouglou** [comment]
@player1-github thanks for your feedback ! really appreciate ! Will follow this PR then.

**Comment by justinfx** [comment]
@player1-github my last reply was 10 months ago, after the proposed fix was merged. I had tested my reported issue using the latest commits. It would be great if someone could confirm my report on the Yugabyte issue.

**Comment by PierreF** [comment]
This issue looks like the same I had: it happen if the initial set of Cassandra IP addresses change.

This happen easily on Kubernetes. With a rolling upgrade of Cassandra, each Cassandra node will be restarted with a new IP address. When the last node is restarted all initial IPs have changed and gocql fail with "no hosts available in the pool"

I've create a minimal case to reproduce this issue in the following repository using Docker compose: https://github.com/PierreF/gocql-cluster-issue

This repository has:
* A Cassandra cluster of 3 nodes.
* It restart one Cassandra after the other. So at least two Cassandra are always up.
* Python client that have no issue. So this rolling restart should be valid.

I've found some other issue that looks duplicate of this issue:
* https://github.com/gocql/gocql/issues/831
* https://github.com/gocql/gocql/issues/1582
* https://github.com/gocql/gocql/issues/1582 (which contains step to reproduce the issue using minikube)
* https://github.com/gocql/gocql/issues/1667
* https://github.com/gocql/gocql/pull/1669 (this PR does not fix my issue)

**Comment by PierreF** [comment]
Inspired by the [following comment](https://github.com/gocql/gocql/issues/831#issuecomment-1325855559), I've tested a workaround which fix my issue: https://github.com/PierreF/gocql-cluster-issue/commit/a7a6db25d8d864eb065d5774528e329ef1dc38e2

**Comment by justinfx** [comment]
@PierreF that workaround looks like a simplified version of what I commented with a year ago, earlier in this thread: https://github.com/yugabyte/yugabyte-db/issues/10182#issuecomment-993792821

It is pretty ugly and has to be a bit more careful in guarding concurrent access to the session in queries that are inflight. I wouldn't really consider it a great workaround. It's more like a band-aid

**Comment by sseidman** [comment]
@PierreF I'd be curious if using both https://github.com/gocql/gocql/pull/1670 and https://github.com/gocql/gocql/pull/1669 would solve your issue. I've had positive results using both changes as a workaround

**Comment by mfamador** [comment]
> @PierreF I'd be curious if using both #1670 and #1669 would solve your issue. I've had positive results using both changes as a workaround

I've used both changes in those PRs and they didn't solve my issue, at least the way I'm testing it. 
We've been hitting this issue often on Azure CosmosDB with Cassandra interface, and i't hard to reproduce it with it.
The way I'm testing it's running a local cassandra container and removing it while my consumer is working (if I just stop it and start it there's no issue, it reconnects immediately). When recreating the container it won't reconnect unless I use the "band-aid" from @PierreF, that is working for me.

**Comment by PierreF** [comment]
If the test-case I've done in https://github.com/gocql/gocql/issues/915#issuecomment-1422870407 (https://github.com/PierreF/gocql-cluster-issue) is the issue @zhixinwen reported, then #1682 fix this issue.

At least after an upgrade to gocql v1.4.0, I no longer have an issue.

**Comment by josuebrunel** [comment]
Still having the same issue with Azure Cosmos DB Cassandra Interface, even after upgrading to v1.5.2

**Comment by amcrn** [comment]
For those who are still experiencing this issue, can you confirm whether you're using `DisableInitialHostLookup` or not? See https://github.com/gocql/gocql/issues/1721

**Comment by josuebrunel** [comment]
I do use `DisableInitialHostLookup=true`

**Comment by dustinblue** [comment]
I'm using `cluster.DisableInitialHostLookup = true` also.  Was getting the error on 1.5.x and 1.6.x using aws keyspaces.  Went back to `github.com/gocql/gocql v0.0.0-20200410100145-b454769479c6` haven't had any issues since.

**Comment by martin-sucha** [comment]
We are considering deprecation of `DisableInitialHostLookup=true` (see comments in https://github.com/gocql/gocql/issues/1721). Is using `SingleHostReadyPolicy`+`DisableInitialHostLookup=false` good enough for those who are using `DisableInitialHostLookup` now? It will achieve shorter time to startup the session too. Are there any other use cases besides the lower session startup time?

**Comment by josuebrunel** [comment]
Hello @jack-at-circle 
I had the same issue with Azure and the following cluster' config helped me:
```go
cluster.DisableInitialHostLookup=false
cluster.PoolConfig = gocql.PoolConfig{
                HostSelectionPolicy: gocql.SingleHostReadyPolicy(gocql.RoundRobinHostPolicy()),
 }
```

Hope it helps

**Comment by jack-at-circle** [comment]
We're currently using that config @josuebrunel but are still losing connections after a day or two : (

**Comment by josuebrunel** [comment]
@jack-at-circle Oh I see. I'm using `github.com/gocql/gocql v1.5.2` and I also have a *retry policy* (Can't really tell if it has an effect or not). Something like:

```go
type ReconnectionPolicy struct {
        maxRetries int
        delay      time.Duration
}

func (rp ReconnectionPolicy) GetInterval(currentRetry int) time.Duration {
        slog.Info("Trying to reconnect to db instance", "currentRetry", currentRetry)
        return rp.delay
}

func (rp ReconnectionPolicy) GetMaxRetries() int {
        return rp.maxRetries
}


cluster.ReconnectionPolicy = ReconnectionPolicy{maxRetries: 4, delay: time.Duration(400) * time.Millisecond}
```

**Comment by Rikkuru** [comment]
I had this problem with v1.6.0 (actually replace github.com/gocql/gocql => github.com/scylladb/gocql v1.12.0)

It was a configuration isuue. (thankfully only in dev environment)

My nodes have 2 ip addrs 100.* (extip) and 10.*(ip). Config made use of extip  addrs but system.local returns ip addrs. And firewall only allowed extip addrs
Service was able to start as initial addr was taken from config, but all the other nodes used ip provided by scylla and were down.
If this one node went down too we ended up without conns to scylla 
There is a fallback to cfg addrs in control but when it eventually succeeded setupConn still used this conn to request system.local  and added received ips to ring.

Log
```
2024/05/23 14:38:20.717959 Session.ring:[100.*:UP][10.*:DOWN]
// after 100.* went down 
2024/05/23 15:16:20.717901 Session.ring:[10.*:DOWN][10.*:DOWN] 
// 100.* returned there are attemts to fallback to config but ring is not changed (all errors on dial to 10.* addrs )
2024/05/23 15:22:16.845645 gocql: control falling back to initial contact points.
```

**Comment by djschaap** [comment]
This issue is seven years old as of yesterday :birthday: - quite a few changes have been made to the connection failure recovery logic since it was created.

Might I suggest closing this issue and documenting each remaining failure case in its own (new) issue?

**Comment by bhuvi-maheshwari** [comment]
We are still facing this issue. Restarting the client fixes it. Otherwise every other day we see same error. What is the fix for this problem?

**Comment by djschaap** [comment]
@bhuvi-maheshwari  This issue has been closed for months.  If you've confirmed your configuration is correct and suspect a bug, I'd suggest opening a new issue, including all versions, configuration, and log messages specific to your situation.

**Comment by dkropachev** [comment]
> We are still facing this issue. Restarting the client fixes it. Otherwise every other day we see same error. What is the fix for this problem?

As issue suggests problem is that `broadcast_rpc_address` of the nodes are not accessible from the client, there is no way to fix that, unless you implement `AddressTranslator` or fix `broadcast_rpc_address` that will make it accessible.

**Comment by bhuvi-maheshwari** [comment]
@dkropachev Thanks for your reply. Then how do we fix it? Do you know if upgrading the version of gocql will fix it? If yes then what version? We are currently on v1.6.0

**Comment by dkropachev** [comment]
@bhuvi-maheshwari , to answer that question I need to know you network infrastructure and driver configuration.

**Comment by bhuvi-maheshwari** [comment]
@dkropachev Here is the cassandra setup. We use github.com/gocql/gocql v1.6.0 version. Go version: 1.21.1

clusterConfig := gocql.NewCluster(ips...)
clusterConfig.Port = port
clusterConfig.Keyspace = keyspace
clusterConfig.SslOpts = &gocql.SslOptions{CaPath: certpath}
clusterConfig.Consistency = gocql.LocalQuorum
clusterConfig.Authenticator = gocql.PasswordAuthenticator{
	Username: uname,
	Password: password,
}
clusterConfig.ProtoVersion = 4
// Set retry policy to retry 3 times in case of transient failures
// clusterConfig.RetryPolicy = &gocql.SimpleRetryPolicy{NumRetries: 3}
clusterConfig.ReconnectInterval = time.Duration(interval) * time.Second
clusterConfig.Timeout = time.Duration(timeout) * time.Second
clusterConfig.DisableInitialHostLookup = false
clusterConfig.ReconnectionPolicy = ReconnectionPolicy{maxRetries: 4, delay: time.Duration(400) * time.Millisecond}
s := new(StorageRepository)
s.ClusterConfig = clusterConfig
if s.Session, err = clusterConfig.CreateSession(); err != nil {
	log.Fatalf("cassandra.NewStorage => DATABASE_CONNECT_ERROR: %s", err.Error())
}

**Comment by george-rogers** [comment]
This issue persists with the latest version (v1.7.0)

**Comment by sinankivanc3131** [comment]
could you confirm if you are still getting the same problem still @george-rogers  ?

asking because we are also getting this problem

**Comment by george-rogers** [comment]
@sinankivanc3131 yes, we are. The workaround so far has been very suboptimal: reconnect to the service whenever this error is detected, which means a delay of around 30s for us.

**Comment by sinankivanc3131** [comment]
@george-rogers  yes we also fix it by restarting the kubernetes pod to create connections from scratch, is there a easy way to trigger this bug?  We want to try reproducing by restarting cassandra pods manually.

**Comment by t101804** [comment]
> [@sinankivanc3131](https://github.com/sinankivanc3131) yes, we are. The workaround so far has been very suboptimal: reconnect to the service whenever this error is detected, which means a delay of around 30s for us.

can u give ur retry logic, i've countering this issue also
